### PR TITLE
INT-5201 Batch Application Failures

### PR DESCRIPTION
**Pull request by edreyer1**
The error in this ticket [https://jupiterone.atlassian.net/browse/INT-5201](url) occurs when you try to list the batch applications in an account that is not linked to an auto-storage account. You are unable to create an application if the account is currently unlink to an auto-storage account but the account can be unlinked after application is created.

Steps I did to setup azure to produce this error: 

1. Create Batch Account that is linked to an auto-storage account
2. Create Batch Application
3. Select "None" when selecting a storage account under the "Storage account" tab

This adds a check that autoStorage (defined in rawData of batch account) is defined before iterating through batch applications for that account

### CODE REVIEW

File: src/steps/resource-manager/batch/index.ts [hunk 1]
```diff
@@ -120,28 +122,31 @@ export async function fetchBatchApplications(
...
-              _class: RelationshipClass.HAS,
-              from: batchAccountEntity,
-              to: batchApplicationEntity,
-            }),
-          );
-        },
-      );
+      const rawData = getRawData<BatchAccount>(batchAccountEntity);
+
+      if (rawData?.autoStorage) {
```
Comment (ndowmon) [inline src/steps/resource-manager/batch/index.ts:127]: Excellent write-up in the PR description! I had a few follow-up questions on this point:

> You are unable to create an application if the account is currently unlink to an auto-storage account but the account can be unlinked after application is created.

1. As soon as the auto-storage account is unlinked, does this API return a `409` again?
2. What happens to the batch application after the auto-storage account is unlinked? Is it deleted, or does it stick around?
3. If the batch application is _not_ deleted after the auto-storage account is unlinked, are we still ingesting it somehow? Or is it simply ignored? Is there any way to actually ingest it in this case?

File: src/steps/resource-manager/batch/index.ts [hunk 2]
```diff
@@ -120,28 +122,31 @@ export async function fetchBatchApplications(
...
-          await jobState.addRelationship(
-            createDirectRelationship({
-              _class: RelationshipClass.HAS,
-              from: batchAccountEntity,
-              to: batchApplicationEntity,
-            }),
-          );
-        },
-      );
+      const rawData = getRawData<BatchAccount>(batchAccountEntity);
```
Comment (ndowmon) [inline src/steps/resource-manager/batch/index.ts:125]: nit: use `batchAccountEntity` and `batchAccount` as variable names, as described in the **Fetch Child Entities** step pattern documentation https://github.com/JupiterOne/sdk/blob/main/docs/integrations/step-patterns.md#fetch-child-entities-step-pattern

```suggestion
      const batchAccount = getRawData<BatchAccount>(batchAccountEntity);
```

File: src/steps/resource-manager/batch/index.test.ts [hunk 3]
```diff
@@ -195,6 +195,30 @@ describe('step - batch applications', () => {
       _type: 'azure_batch_account',
       _class: ['Service'],
       name: 'j1devbatchaccount',
+      _rawData: [
```
Comment (ndowmon) [inline src/steps/resource-manager/batch/index.test.ts:198]: I'd like to start replacing these legacy, hard-coded tests with tests that use `executeStepWithDependencies`. In this project, there is a test at `src/steps/resource-manager/storage/index.test.ts` that uses the newer patterns - can you use that as a reference to update the test for the **Batch Application** step?

```ts
test('rm-storage-containers', async () => {
  const stepTestConfig = getStepTestConfigForStep(steps.STORAGE_CONTAINERS);

  recording = setupAzureRecording({
    name: 'rm-storage-containers',
    directory: __dirname,
    options: {
      matchRequestsBy: getMatchRequestsBy({
        config: stepTestConfig.instanceConfig,
      }),
    },
  });

  const stepResults = await executeStepWithDependencies(stepTestConfig);
  expect(stepResults).toMatchStepMetadata(stepTestConfig);
}, 10_000);
```

**Review by ndowmon — DISMISSED** [review]
Looks good! I've approved, but I also made some suggestions - I think small changes like this are good opportunities to improve our testing in this project, which is in need of refinement.

File: src/steps/resource-manager/batch/index.ts (hunk 1 above)
Comment (edreyer1) [reply src/steps/resource-manager/batch/index.ts:127]: As soon as the auto-storage account is unlinked  we start getting the 409 errors again. The application doesn't get deleted, if you relink the auto-storage account you are able to start ingest information about the applications again

File: src/steps/resource-manager/batch/index.test.ts [hunk 4]
```diff
@@ -1,18 +1,23 @@
 import {
   fetchBatchAccounts,
-  fetchBatchApplications,
+  //fetchBatchApplications,
```
Comment (ndowmon) [inline src/steps/resource-manager/batch/index.test.ts:3]:
```suggestion
```

**Review by ndowmon — APPROVED** [review]

**Pull request merged**